[#57529] Add options for sum totals calculation mode in hierarchies

aaron-contreras opened the pull request:
# Ticket
https://community.openproject.org/work_packages/57529

# What are you trying to accomplish?
Adding a new setting in order to allow users to determine the way in which they'd like sum totals of percent complete in work package hierarchies to be calculated.

1. **Simple average (default)**: Work is ignored and the total % Complete will be a simple average of % Complete values of work packages in the hierarchy.
2. **Weighted by work**: The total % Complete will be weighted against the Work of each work package in the hierarchy. Work packages without Work will be ignored.

## Screenshots
Not yet

# What approach did you choose and why?
1. Extend the `UpdateAncestorsService` to calculate the total % complete based on the current Setting value.
2. Add a radio group entry to the progress tracking settings page to be able to toggle between the modes.
3. Add a job that recalculates all total % complete values that gets triggered when the mode is changed in the settings page.

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

aaron-contreras committed:
Create setting for Calculation of % Complete hierarchy totals

* Adds the settings to the definitions file.

* Sets the default and allowed values.

aaron-contreras committed:
Incorporate simple average mode into total % complete calculations

* Adds setting into derivation code.

* Adds specs for the simple average mode calculation.

aaron-contreras committed:
Adjust selection of done ratio based on new rules for children

aaron-contreras committed:
Add radio group to progress tracking form

Now the settings can be changed from the UI.

aaron-contreras committed:
Add feature spec for setting toggling

* Fixes cuprite flag to enable cuprite in this spec file.

aaron-contreras committed:
Add recalculation in job when switching to work weighted mode

* Missing to add journal entries

aaron-contreras committed:
Re-order radio group options according to design document

Calculation of % Complete hierarchy totals should be the second
setting on the form

aaron-contreras committed:
Trigger the recalculation job when changing to work weighted mode

* Triggers job
* Adds journal entries with specs and translations

aaron-contreras committed:
Add in simple average calculation mode job

* Adds temporary depth table - Could be cleaned up down thie line with
  the `work_package_hierarchies` table.

cbliard commented on spec/features/admin/progress_tracking_spec.rb:
```diff
@@ -28,7 +28,9 @@
-RSpec.describe "Progress tracking admin page", :cuprite, :js,
+RSpec.describe "Progress tracking admin page",
+               :js,
+               :with_cuprite,
```
I was not lucky, I copied it over from the only spec that uses `:cuprite`: `spec/features/admin/languages_settings_spec.rb`. I thought that it had been renamed at some point.

aaron-contreras committed:
Account for status exclusion from total calculations in simple average mode

cbliard commented on app/services/settings/update_service.rb:
```diff
@@ -47,6 +47,10 @@ def set_setting_value(name, value)
+    elsif name == :total_percent_complete_mode && old_value != "work_weighted_average" && new_value == "work_weighted_average"
+      WorkPackages::Progress::ApplyTotalPercentCompleteModeChangeJob
+        .perform_later(mode: new_value,
+                       cause_type: "total_percent_complete_mode_changed_to_work_weighted_average")
```
Does it mean that both "Progress calculation mode" and "total % complete calculation mode" can't be changed at the same time?

cbliard replied:
Ah, I did not think enough for this one. It's called once for each modified setting, so it will be called for both and 2 jobs will be triggered.
It's probably not ideal to have the journal being updated twice, but that will do for now.

cbliard reviewed (commented):
Nice work so far 👏 

I only looked at code so far. I need to play a bit with it now.

cbliard commented on spec/services/work_packages/update_ancestors_service_spec.rb:
```diff
@@ -795,6 +796,213 @@ def call_update_ancestors_service(work_package)
 
+  describe "simple average mode for total % complete calculation",
+           with_settings: { total_percent_complete_mode: "simple_average" } do
+    subject(:call_result) do
+      described_class.new(user:, work_package: parent)
+                      .call(%i(remaining_hours))
+    end
```
That's not exactly how this class is supposed to be called: it is called on a child which has been modified, and we check that the ancestors got updated as well.

Calling it like you did actually hides a bug: in the `WorkPackages::UpdateAncestors::Loader`, there can be structured data (the `WorkPackageLikeStruct`), but also plain `WorkPackage` instances, and they not respond to `status_excluded_from_totals`.

I'll fix that.

cbliard replied:
Fixed in 1874fb3003e.

cbliard committed:
Fix `status_excluded_from_totals` not existing for `WorkPackage`

In `WorkPackages::UpdateAncestors::Loader`, the returned elements are
not always `WorkPackageLikeStruct` instances: whenever possible the
loader reuses knows instances of `WorkPackage` when looking for
ancestors or descendants.

cbliard commented on app/services/work_packages/update_ancestors_service.rb:
```diff
@@ -120,15 +120,43 @@ def derive_done_ratio(ancestor, loader)
+      all_done_ratios << work_package.done_ratio
+    end
+
+    progress = all_done_ratios.sum.to_f / all_done_ratios.count
+    progress.round
```
Found another issue: this fails if all work packages are excluded: the sum returns 0.0, count returns 0, so progress will be `NaN`, and `NaN.round` raises.

In this case, ∑ % complete should be cleared.

I'll fix that.

cbliard replied:
Fixed in f0408482463

cbliard committed:
Fix when all work packages of a hierarchy are excluded from totals

In this case, there is no calculated simple average: the ∑ % Complete
should be cleared.

cbliard committed:
Removed flaky spec

I'm not sure why it fails now while it did not before. Maybe it's when
switching to cuprite. Anyway this test is not worth enough to bother
fixing it.

cbliard committed:
Fix rubocop warnings Metrics/AbcSize and Metrics/PerceivedComplexity

cbliard committed:
Add a test when all hierarchy is excluded from totals

Just to be extra-sure this case is handled correctly in jobs as well.

cbliard committed:
Add failing spec about handling simple average mode in a job

cbliard committed:
Use work_package_hierarchies to create depth table

Simpler.

cbliard committed:
Compute total % complete according to mode in update status job

Reuse and duplicate a lot of logic from the other job. It needs to be
reworked, but it works

aaron-contreras committed:
Merge branch 'dev' into implementation/57529-options-for-total-calculation-sum-mode

aaron-contreras committed:
Add extra spec specifically accounting for excluded statuses

aaron-contreras committed:
Test for unset total values

aaron-contreras committed:
Remove unnecessary update

aaron-contreras committed:
Extract common operations to sql_commands

cbliard commented on app/workers/work_packages/progress/sql_commands.rb:
```diff
@@ -188,7 +188,6 @@ def update_totals
     if Setting.total_percent_complete_mode == "work_weighted_average"
       update_total_percent_complete_in_work_weighted_average_mode
     elsif Setting.total_percent_complete_mode == "simple_average"
-      update_total_percent_complete_in_work_weighted_average_mode
```
Nice!

cbliard reviewed (approved):
Well done. I think it's good to go now.
